### ISSUE TITLE
Don't make redundant copies of the same module package.

### ISSUE THREAD

**Issue by joeaawad** [opentofu/opentofu#1086]
### Summary

Currently, OpenTofu can make thousands of copies of a single module package, which can be very expensive, both in terms of time to make so many copies and in disk usage. This could be refactored to substantially speed up inits for many users.

### Problem Statement

Currently, when using remote module sources (for this example I will use GitHub source, but it applies equally to all types of remote module sources), OpenTofu makes separate copies of the entire module package for every single module source on disk. This means that if I have 100 different module calls that specify anywhere between 1 and 100 different [module source sub-directories from a single module package](https://opentofu.org/docs/language/modules/sources/#modules-in-package-sub-directories), the behavior we will see is that it will download the module package once from the remote source, then make 99 more local copies of that module package. 

This behavior can seem relatively minor, but several types of common OpenTofu usage can make this behavior very painful. First, if a company has a single module package (such as a git repo) where they store all of their shared modules in separate subdirectories that numerous internal teams working in numerous separate repos are able to reference via remote module source. This can make the single module package quite large, but avoids having hundreds of separate tiny repos if you have hundreds of shared modules. Second, if you either have a single large root module that references scores or hundreds of remote modules, or numerous root modules that are run in a since automated job that collectively have hundreds or thousands of module invocations. In these cases, profiling the job can identify that just the local copying of a single module package that is being used thousands of times can take upwards of 40-50% of the total execution time and fill up a ton of disk space. 

This is also inconsistent with the behavior when using a local path for the module source. When using local path, the init simply provides a pointer to the existing module package on disk, it does not make any copies of the module package. This can result in meaningful differences if someone is using `path.module` to write files to disk, as noted in related issues.

### User-facing description

The user facing impact is:

1. Reduced execution time when using a single module package multiple times in a given root module
2. Reduced disk usage because the module package is not copied multiple times

### Technical Description

See related issues, but TLDR is changing the behavior of Tofu init so that instead of copying the initially downloaded module package on disk, OpenTofu just creates a pointer to the initial module package download for subsequent usages.

From a technical implementation standpoint, it should be a fairly trivial change since you are just changing the functionality when OpenTofu sees that it is init-ing a single module package for not the first time so that, instead of the existing behavior of copying the full module package, it invokes the existing functionality that it already uses for local module paths to instead point to the existing module package. The current code to copy a module package could probably be removed unless if it is invoked elsewhere.

### Rationale and alternatives

This feature would provide a meaningful performance improvement for power users such as my company, since we have a number of very long running jobs that can take up to as much as an hour to plan a bunch of different Terraform root modules. Having that time get reduced by 40-50% would be a huge reason to cut over to OpenTofu.

Alternatives include not using remote module sources and instead using local module sources with a pre-init step to download and populate a cache with the remote module package at a known path that the local paths point to. This is completely viable, but not as nice of a user experience and requires custom tool that all users at the company would have to use.

### Downsides

Technically, this could be considered a breaking change (see related issues). However, my argument is that this should NOT be considered a breaking change because there is currently inconsistent between behavior when using a local module source and a remote module source. This inconsistency has never been documented or recommended and per prior documentation, OpenTofu should always behave exactly the same under the hood regardless of where a module is sourced from. As a result, I see this as a bugfix even if it will break a small percentage of users who are, in my opinion, improperly using `path.module` for write operations. Also, this could be something that is part of a living document of differences between Terraform and OpenTofu as part of a migration guide so that people can make the relatively minor change of using `path.root` instead of existing usages of `path.module` for existing use cases when they migrate.

### Related Issues

- https://github.com/hashicorp/terraform/issues/29503
- https://github.com/hashicorp/terraform/issues/31422

**Comment by cam72cam** [comment]
For clarity, links to TF's issues as reference is fine.  We just need the description here.

**Comment by joeaawad** [comment]
Updated per feedback about referencing TF issues. LMK if those changes are acceptable and if there is any additional information I can provide to push along consideration of this.

**Comment by cam72cam** [comment]
I'm a huge fan of this issue, it's caused lots of sitting around and waiting for no reason.

**Comment by cam72cam** [comment]
I agree that the remote functionality differing from the local functionality is probably in the classification of a bug, but it's possible that people depend upon it.

I think that a experimental flag or env variable to enable this functionality, with a deprecation warning in the documentation + release notes is the way I would approach it.  That would give people time to check their modules and keep the old functionality if need be when the default behavior eventually changes.  I'm curious what the other @opentofu/core-engineers think?

**Comment by joeaawad** [comment]
@cube2222 @cam72cam I created [this test case](https://github.com/joeaawad/opentofu-large-module-source) so that you can easily observe the behavior yourselves. Please let me know if that makes sense or you have any questions.

To provide a bit more background, my company's centralized module package is ~35MB with ~3k files. For context, this sample playground is 8.7MB with ~1k files so substantially smaller, but still large enough to clearly demonstrate the issue. In my testing, the change I am requesting in this issue would decrease `init` times by more than 75% for my use case, which based on how long our total `init` and `plan` runs take and what percentage of that time is spent on the `init` portion, would reduce the total runtime by roughly 40-50% on average. 

If OpenTofu chose to make this change, it would not only encourage my company to switch to OpenTofu, but would also save companies like Spacelife, env0, Scalr, and Harness that run managed OpenTofu pipelines for customers substantial compute costs because the init would occur so much faster for quite a few customers who have a single module package used multiple times.

**Comment by cam72cam** [comment]
I hacked the codebase to see what copying in parallel could do:
```
original
init 0m7.929s
initx2 0m0.055s

go func
init 0m1.965s
initx2 0m0.056s
```

It's a pretty decent improvement with lots of large modules and matches your README.  However, it ran into some interesting parallel edge cases that would require more thought and is not a completely trivial change.

I feel your frustration and agree that the current functionality of remote modules vs local modules was not well thought out in the `module.path` context.  The problem is figuring out how to make this sort of change without blowing up a lot of existing modules in a unexpected or hard to detect way.

The one thing I don't fully understand is your assertion that it would improve plan times as well.  Plan just checks that the modules already exist.  Is it possible that your plan workflow is actually a HCL parse bottleneck?

**Comment by joeaawad** [comment]
That's good to see that we could potentially get some improvement with parallelization although it sounds like there are some possible land mines in that process. If that is the only option OpenTofu is open to considering, I would certainly welcome it and encourage its rapid implementation. 

However, I still think that the best solution would be the change to avoid duplicate copies of the module packages in the first place, even if it is currently completed as opt-in functionality (that you could potentially make the default be opt-in whenever there is a 2.0) that can be set in either the [config file](https://opentofu.org/docs/cli/config/config-file) or an [environment variable](https://opentofu.org/docs/cli/config/environment-variables) like `TF_SEPARATE_MODULE_PACKAGES=FALSE`  or similar depending on how you want to word it. This would not only be substantially faster than anything you are able to do with increasing copy parallelization, but also make a huge difference for disk usage, which can result in huge performance boosts in situations like 1) already parallelizing multiple Terraform runs in different directories on a single machine 2) having security scanning tooling that negatively impacts performance, particularly when it sees such large scale file copying.

**Comment by cam72cam** [comment]
Part of the problem is that path.module can be passed to scripts and providers which we can't track.

I think that this would have to be behind a feature flag and/or be a deliberate breaking change.

**Comment by cam72cam** [comment]
@joeaawad there's an interesting point here about for_each and count.  As far as I can tell, in the remote case they *share* the directory.

So even for remote, it depends entirely on how it's being used.  In my mind that backs up your argument to unify it, but it's still a breaking change.

**Comment by cam72cam** [comment]
root -> moduleA(for_each) -> moduleB -> moduleC

Means that all moduleBs and moduleCs will share the same directory, even if it's remote

**Comment by skyzyx** [comment]
Sounds related-to/duplicative-of https://github.com/opentofu/opentofu/issues/1199.

**Comment by bysiber** [comment]
The module duplication problem is significant at scale. When you have a module referenced multiple times in a configuration (different paths/versions), Terraform/OpenTofu copies the entire module package each time, leading to massive `.terraform/` directories.

On macOS, with multiple Terraform projects, the `.terraform/` directories combined can consume 5-15GB -- and that's before accounting for the provider plugin cache at `~/.terraform.d/plugin-cache/`:

```bash
# Check total .terraform usage across projects
find ~/Projects -name ".terraform" -type d -maxdepth 4 -exec du -sh {} + 2>/dev/null | sort -rh | head -15

# Check provider plugin cache
du -sh ~/.terraform.d/plugin-cache/ 2>/dev/null

# Check total provider binaries
du -sh ~/.terraform.d/ 2>/dev/null
```

Deduplicating module copies (via symlinks or content-addressed storage) would cut the storage overhead significantly. Until then, periodic cleanup of stale project `.terraform/` directories is the workaround.

For macOS developers: [ClearDisk](https://github.com/bysiber/cleardisk) is a free, open-source menu bar app that monitors developer cache directories including Terraform/OpenTofu alongside Docker, Xcode DerivedData, node_modules, Rust target/, Go modules, Homebrew, and more. Gives a quick overview of what's eating your SSD.

**Comment by apparentlymart** [comment]
Hi @bysiber,

The analysis you described there seems like it would include both provider packages _and_ module packages, making it broader than what this issue is about.

To measure only copies of modules -- which is what this issue is focused on -- you'd need to limit only to the `.terraform/modules` subdirectory. I'd be curious to know what the storage footprint is of _just_ the cached module packages.

The storage size of cached provider packages is also potentially large of course, but that's beyond the scope of this issue. The system already creates symlinks to the global provider cache directory when possible, so if that doesn't appear to be working for you then we could talk about that in a separate issue from this one, or in [Discussions](https://github.com/orgs/opentofu/discussions).

**Comment by bysiber** [comment]
@apparentlymart Makes sense, focusing on just .terraform/modules gives a much cleaner picture for this issue. The module dirs are smaller individually compared to providers yeah, but across a bunch of projects they still add up, especially when the same modules keep getting duplicated in each init.

Good to know the provider cache already uses symlinks from the global dir.

**Comment by jessefarinacci** [comment]
Here's some real world data from a smallish, mostly "serverless", dumb/static, web-based project; maybe it is too much noise to signal, my apologies if this isn't useful. If anyone has some better ideas to report statistical information anonymously I will listen and try to provide..

My setup relies /heavily/ on reusable modules, often with multiple layers of nesting. The top level are resolved via `git::codecommit://` from one giant repository/module monolith, with internal module references via local file (e.g. `module { source = ../s3/bucket }` style) though I'd like to break it up in future to support isolated testing, linting, etc.

In the output below, everything is initialized and installed, no drift, no statistical tricks, blah blah. It's my working state for performing meaningful tasks in the project.

```
❯ tofu version
OpenTofu v1.11.5
on darwin_arm64
+ provider registry.opentofu.org/hashicorp/archive v2.7.1
+ provider registry.opentofu.org/hashicorp/aws v6.33.0
+ provider registry.opentofu.org/hashicorp/awscc v1.72.0
+ provider registry.opentofu.org/hashicorp/local v2.7.0
+ provider registry.opentofu.org/hashicorp/null v3.2.4
+ provider registry.opentofu.org/hashicorp/random v3.8.1
```

```
❯ du -d2 -h
  0B	./.terraform/providers
152M	./.terraform/modules
153M	./.terraform
4.0K	./modules
153M	.
```

```
❯ ls *.tf | wc -l
111

❯ grep '^data' *.tf | wc -l
26

❯ grep '^resource' *.tf | wc -l
0

❯ grep '^module' *.tf | wc -l
80

❯ jq --raw-output '.resources[] | "\(.mode).\(.type)"' terraform.tfstate | sort | uniq -c | sort -n | tac
 151 data.aws_ssm_parameter
  77 managed.aws_ssm_parameter
  59 managed.aws_route53_record
  55 managed.aws_iam_service_linked_role
  54 data.aws_servicequotas_service_quota
  29 managed.random_uuid
  25 managed.aws_iam_role_policy_attachment
  24 data.aws_iam_role
  23 managed.aws_iam_policy
  22 managed.aws_s3_bucket_versioning
  22 managed.aws_s3_bucket_server_side_encryption_configuration
  22 managed.aws_s3_bucket_request_payment_configuration
  22 managed.aws_s3_bucket_replication_configuration
  22 managed.aws_s3_bucket_public_access_block
  22 managed.aws_s3_bucket_policy
  22 managed.aws_s3_bucket_ownership_controls
  22 managed.aws_s3_bucket_notification
  22 managed.aws_s3_bucket_lifecycle_configuration
  22 managed.aws_s3_bucket_inventory
  22 managed.aws_s3_bucket_intelligent_tiering_configuration
  22 managed.aws_s3_bucket_accelerate_configuration
  22 managed.aws_s3_bucket
  22 managed.aws_cloudwatch_log_group
  21 managed.aws_cloudwatch_event_target
  21 managed.aws_cloudwatch_event_rule
  19 data.aws_route53_zone
  15 managed.aws_codecommit_repository
  14 managed.aws_sesv2_email_identity
  13 data.aws_iam_policy
  12 managed.aws_ses_receipt_rule
  12 managed.aws_iam_role
  11 managed.aws_iam_user
  11 managed.aws_iam_group
  11 data.aws_cloudfront_response_headers_policy
  11 data.aws_cloudfront_origin_request_policy
  11 data.aws_cloudfront_cache_policy
  10 managed.aws_iam_virtual_mfa_device
  10 managed.aws_iam_group_policy_attachments_exclusive
  10 managed.aws_iam_group_membership
  10 managed.aws_iam_access_key
   8 managed.aws_cloudfront_function
   7 managed.aws_acm_certificate_validation
   7 managed.aws_acm_certificate
   6 managed.random_string
   6 managed.aws_macie2_findings_filter
   6 managed.aws_cloudwatch_log_delivery_source
   6 managed.aws_cloudwatch_log_delivery_destination
   6 managed.aws_cloudwatch_log_delivery
   6 managed.aws_cloudfrontkeyvaluestore_key
   6 managed.aws_cloudfront_distribution
   6 data.aws_regions
   6 data.aws_cloudfront_function
   5 managed.aws_resourceexplorer2_view
   5 managed.aws_cognito_user_pool_client
   5 managed.aws_cloudwatch_metric_alarm
   5 managed.aws_cloudfront_key_value_store
   5 data.aws_s3_bucket
   4 managed.aws_ssm_service_setting
   4 managed.aws_rum_app_monitor
   4 managed.aws_cloudfront_response_headers_policy
   4 managed.aws_cloudfront_origin_request_policy
   4 managed.aws_cloudfront_origin_access_control
   4 managed.aws_cloudfront_cache_policy
   4 data.archive_file
   3 managed.null_resource
   3 managed.aws_sesv2_email_identity_mail_from_attributes
   3 managed.aws_sesv2_email_identity_feedback_attributes
   3 managed.aws_sesv2_configuration_set_event_destination
   3 managed.aws_sesv2_configuration_set
   3 managed.aws_s3_object
   3 managed.aws_ecr_account_setting
   3 managed.aws_codeartifact_repository
   3 managed.aws_cloudwatch_log_resource_policy
   3 data.aws_partition
   3 data.aws_organizations_organization
   3 data.aws_codeartifact_repository_endpoint
   3 data.aws_cloudwatch_event_bus
   3 data.aws_caller_identity
   2 managed.aws_sqs_queue
   2 managed.aws_sns_topic_policy
   2 managed.aws_sns_topic
   2 managed.aws_ses_receipt_rule_set
   2 managed.aws_securityhub_insight
   2 managed.aws_lambda_layer_version
   2 managed.aws_kms_replica_key
   2 managed.aws_kms_key_policy
   2 managed.aws_kms_key
   2 managed.aws_iam_group_policy_attachment
   2 managed.aws_dynamodb_table
   2 managed.aws_cognito_user
   2 data.aws_kms_key
   2 data.aws_dynamodb_table
   1 managed.aws_xray_resource_policy
   1 managed.aws_xray_encryption_config
   1 managed.aws_wafv2_web_acl_logging_configuration
   1 managed.aws_wafv2_web_acl
   1 managed.aws_synthetics_canary
   1 managed.aws_sqs_queue_redrive_policy
   1 managed.aws_sqs_queue_redrive_allow_policy
   1 managed.aws_sqs_queue_policy
   1 managed.aws_sns_topic_subscription
   1 managed.aws_sfn_state_machine
   1 managed.aws_sesv2_contact_list
   1 managed.aws_sesv2_account_suppression_attributes
   1 managed.aws_ses_active_receipt_rule_set
   1 managed.aws_securityhub_account
   1 managed.aws_s3_account_public_access_block
   1 managed.aws_route53domains_registered_domain
   1 managed.aws_route53domains_delegation_signer_record
   1 managed.aws_route53_zone
   1 managed.aws_route53_query_log
   1 managed.aws_route53_key_signing_key
   1 managed.aws_route53_hosted_zone_dnssec
   1 managed.aws_resourceexplorer2_index
   1 managed.aws_pipes_pipe
   1 managed.aws_oam_link
   1 managed.aws_memorydb_acl
   1 managed.aws_macie2_classification_export_configuration
   1 managed.aws_macie2_account
   1 managed.aws_lambda_runtime_management_config
   1 managed.aws_lambda_permission
   1 managed.aws_lambda_function_recursion_config
   1 managed.aws_lambda_function
   1 managed.aws_iam_user_ssh_key
   1 managed.aws_iam_user_group_membership
   1 managed.aws_iam_service_specific_credential
   1 managed.aws_iam_account_password_policy
   1 managed.aws_iam_account_alias
   1 managed.aws_guardduty_detector
   1 managed.aws_elasticache_user
   1 managed.aws_ecs_account_setting_default
   1 managed.aws_ecr_repository_policy
   1 managed.aws_ecr_repository
   1 managed.aws_ecr_replication_configuration
   1 managed.aws_ecr_registry_scanning_configuration
   1 managed.aws_ecr_registry_policy
   1 managed.aws_ecr_lifecycle_policy
   1 managed.aws_ebs_encryption_by_default
   1 managed.aws_config_retention_configuration
   1 managed.aws_config_delivery_channel
   1 managed.aws_config_configuration_recorder_status
   1 managed.aws_config_configuration_recorder
   1 managed.aws_cognito_user_pool_domain
   1 managed.aws_cognito_user_pool
   1 managed.aws_cognito_risk_configuration
   1 managed.aws_cognito_managed_login_branding
   1 managed.aws_cognito_log_delivery_configuration
   1 managed.aws_codestarconnections_connection
   1 managed.aws_codeartifact_domain
   1 managed.aws_cloudwatch_query_definition
   1 managed.aws_cloudwatch_event_bus
   1 managed.aws_chatbot_slack_channel_configuration
   1 managed.aws_budgets_budget
   1 managed.aws_bedrock_model_invocation_logging_configuration
   1 managed.aws_backup_vault_policy
   1 managed.aws_backup_vault
   1 managed.aws_backup_selection
   1 managed.aws_backup_region_settings
   1 managed.aws_backup_plan
   1 managed.aws_athena_workgroup
   1 managed.aws_api_gateway_account
   1 data.aws_synthetics_runtime_version
   1 data.aws_ses_email_identity
   1 data.aws_s3_object
   1 data.aws_iam_user
```

EDIT: added `grep data *.tf`
EDIT: jq mode and type, thx pete!

**Comment by apparentlymart** [comment]
Hi @diofeher!

Just wanted to check in with you if you are still intending to work on this. No pressure to do it, just that we're trying to finalize what is likely to be completed before the v1.13 feature freeze, and I believe this was still in a research state (figuring out what we ought to do) rather than an implementation state (doing a specific thing) so _probably_ this'll need to get bumped into a later development period now.

I'm going to preemptively shift this into the v1.14.0 milestone for now, but we can put it back in v1.13.0 if we can somehow agree on a plan for how it would get completed in the next week or two.
